src/Microsoft.AspNetCore.Routing/Builder/EndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,189 @@
+            return MapVerbs(builder, pattern, displayName: null, requestDelegate, httpMethods, metadata);
+        }
+
+        public static IEndpointConventionBuilder MapVerbs(
+           this IEndpointRouteBuilder builder,
+           string pattern,
+           string displayName,
+           RequestDelegate requestDelegate,
+           IList<string> httpMethods,
+           params object[] metadata)
+        {
+            if (httpMethods == null)
+            {
+                throw new ArgumentNullException(nameof(httpMethods));
+            }
+
+            List<object> resolvedMetadata = new List<object>();
+            resolvedMetadata.Add(new HttpMethodMetadata(httpMethods));
+            if (metadata != null)
+            {
+                resolvedMetadata.AddRange(metadata);
+            }
+
+            return Map(builder, pattern, displayName ?? $"{pattern} HTTP: {string.Join(", ", httpMethods)}", requestDelegate, metadata: resolvedMetadata.ToArray());
+        }
+        #endregion
+
+        #region Map
+        public static IEndpointConventionBuilder Map(
+            this IEndpointRouteBuilder builder,
+            string pattern,
+            RequestDelegate requestDelegate,
+            params object[] metadata)
+        {
+            return Map(builder, RoutePatternFactory.Parse(pattern), pattern, requestDelegate, metadata);
+        }
+
+        public static IEndpointConventionBuilder Map(
+            this IEndpointRouteBuilder builder,
+            string pattern,
+            string displayName,
+            RequestDelegate requestDelegate,
+            params object[] metadata)
+        {
+            return Map(builder, RoutePatternFactory.Parse(pattern), displayName, requestDelegate, metadata);
+        }
+
+        public static IEndpointConventionBuilder Map(
+            this IEndpointRouteBuilder builder,
+            RoutePattern pattern,
+            RequestDelegate requestDelegate,
+            params object[] metadata)
+        {
+            return Map(builder, pattern, pattern.RawText ?? pattern.DebuggerToString(), requestDelegate, metadata);
+        }
+
+        public static IEndpointConventionBuilder Map(
+            this IEndpointRouteBuilder builder,
+            RoutePattern pattern,
+            string displayName,
+            RequestDelegate requestDelegate,
+            params object[] metadata)
+        {
+            const int defaultOrder = 0;
+
+            var routeEndpointModel = new RouteEndpointModel(
+                requestDelegate,
+                pattern,
+                defaultOrder);
+            routeEndpointModel.DisplayName = displayName;
+            if (metadata != null)
+            {
+                foreach (var item in metadata)
+                {
+                    routeEndpointModel.Metadata.Add(item);
+                }
+            }
+
+            var modelEndpointDataSource = builder.EndpointDataSources.OfType<ModelEndpointDataSource>().FirstOrDefault();
```

Comment:
Storing the data source here instead of in DI. Can do the same for the MvcEndpointDataSource.

Reply:
I think from a scenario point of view.
1. I should be able to resolve **something** from DI taht can find all of the endpoints
2. I should be able to add endpoints via startup
3. I should be able to add endpoints via DI (frameworks)

As long as these aren't violated, that sounds good.

Reply:
Another alternate I see would be for the composite data source to be mutable, and have this API wrap/expose it. Then all of these things are the same.